### DISCUSSION TITLE
Proposal: Centralized Artifact for Quarkus Tooling Checks

### DISCUSSION THREAD

**Discussion by ia3andy** [quarkusio/quarkus#41681] [Design Discussions]
Hi everyone,

We've identified a challenge with ensuring consistent and reliable error checks in Quarkus tooling (IDEs plugins). To address this, I propose creating a centralized artifact (in the `quarkusio` organization for example) containing all checks for Quarkus, MicroProfile (MP), and Qute (and possibly extensions). It could also be three artifacts (one for Quarkus, one for MP and one for Quarkiverse).

#### Objectives:
- **Consistency**: Ensure consistent error messages and reduce false positives.
- **Collaboration**: Facilitate joint efforts in adding, changing, and reviewing checks.
- **Maintenance**: Simplify adherence to specifications for all teams.

#### Structure:
- Each check includes a code, message, and error level.
- Separate lists for Quarkus, MP, and Qute checks.
- Plugins will consume this artifact to apply checks in IDEs.

**====> Result is:** Improved tooling consistency and developer experience.

Your feedback on this proposal is greatly appreciated.

Cheers,

cc @angelozerr @fbricon @maxandersen @cescoffier @gsmet

**Comment by maxandersen** [comment]
When you say artifact do you mean repository or an actual maven artifact that gets used somewhere ?

**Reply by ia3andy** [reply]
They would use the Maven artifact which would contain the checks, here is an example:
https://github.com/redhat-developer/quarkus-ls/blob/master/qute.ls/com.redhat.qute.ls/src/main/java/com/redhat/qute/services/diagnostics/QuteErrorCode.java

**Reply by ia3andy** [reply]
For the Quarkus checks, we would need to extract it, there is no such file yet.

Here is an example of false positive (created by you @maxandersen): https://github.com/redhat-developer/vscode-quarkus/issues/708

**Comment by ia3andy** [comment]
On important aspect is that, the Tooling team should have control on maintenance (releases) so they are not holden by this.

**Comment by maxandersen** [comment]
And what is "each check" ? I feel some context is missing.

**Comment by cescoffier** [comment]
I do not totally understand what you call 'checks'. It is a logged message with an error code? Or is it an exception with in a code/reason m. In both case, we do not use 'code' a lot in Quarkus. So, it might require some effort.

**Comment by maxandersen** [comment]
First off - 1000% agree we should do whatever we can to tools testing that works with Quarkus. 

Second - i don't grok what these "checks" are 

Thirdly - I don't understand why it's important it's in Quarkusio main repo? 

Can you help me grok that ? :)

**Reply by ia3andy** [reply]
Ok, so I initially thought it would be used by the tooling, but now that @cescoffier  says that, most of those checks are also logs in Quarkus. Having a consistant/shared way to deal with those messages (and code?) would be a significant effort, but the result for quality might be worth it.

I initially thought about the Quarkus repo to make sure we take responsibility in reviewing/maintaining them. Using those messages (or whatever we call them) in both might be sensible in the end.

**Reply by maxandersen** [reply]
I honestly is confused...what exactly is it you want to check?

i.e. https://github.com/redhat-developer/vscode-quarkus/issues/708 you is something that would spotted somehow? that issue is because quarkus is more flexible than stock cdi. what would check of "messages" and code help with?

are you talking about having a repository with example code and then run the tooling headless or somehow and check that the list of output warnings/errors are what is expected ?
